docs/design/2020-01-10-shuffle-executor.md

```diff
@@ -0,0 +1,253 @@
+
+When the very position and very moment is found, we inserts a new physical plan `PhysicalShuffle`, which represents the executing plan of Shuffle Executor.
+
+```go
+type PhysicalShuffle struct {
+	basePhysicalPlan
+
+	Concurrency int
+	Tail        PhysicalPlan
+	DataSource  PhysicalPlan
+
+	SplitterType PartitionSplitterType
+	HashByItems  []expression.Expression
+}
+```
+
+For example, when table `t` was created as `create table t (i int, j int)`, the plan of `select j, sum(i) over w from t window w as (partition by j)` would be:
+```
+mysql> explain select j, sum(i) over w from t window w as (partition by j);
++--------------------------+----------+-----------+------------------------------------------------------------+
+| id                       | count    | task      | operator info                                              |
++--------------------------+----------+-----------+------------------------------------------------------------+
+| Projection_7             | 10000.00 | root      | test.t.j, Column#5                                         |
+| └─Window_8               | 10000.00 | root      | sum(cast(test.t.i))->Column#5 over(partition by test.t.j)  |
+|   └─Sort_11              | 10000.00 | root      | test.t.j:asc                                               |
+|     └─TableReader_10     | 10000.00 | root      | data:TableScan_9                                           |
+|       └─TableScan_9      | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
++--------------------------+----------+-----------+------------------------------------------------------------+
+```
+To "shuffle" the Window operator, the plan would changed to:
+```
+mysql> explain select j, sum(i) over w from t window w as (partition by j);
++----------------------------+----------+-----------+------------------------------------------------------------+
+| id                         | count    | task      | operator info                                              |
++----------------------------+----------+-----------+------------------------------------------------------------+
+| Projection_7               | 10000.00 | root      | test.t.j, Column#5                                         |
+| └─Shuffle_12               | 10000.00 | root      | execution info: concurrency:4, data source:TableReader_10  |
+|   └─Window_8               | 10000.00 | root      | sum(cast(test.t.i))->Column#5 over(partition by test.t.j)  |
+|     └─Sort_11              | 10000.00 | root      | test.t.j:asc                                               |
+|       └─TableReader_10     | 10000.00 | root      | data:TableScan_9                                           |
+|         └─TableScan_9      | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
++----------------------------+----------+-----------+------------------------------------------------------------+
+```
```

Comment:
This new plan is pretty strange. Actually in execution, the output of `Shuffle` should be the input of `Window`, so a more intuitive plan from my perspective would be like `TableScan -> TableReader -> Sort -> Shuffle -> Window -> Merge -> Projection`, as you depicted in the diagram above.

Reply:
I agree. I will change this document and related codes in the next couple of days.

Reply:
@eurekaka 
Comment addressed, PTAL~